jupyter_kernel_mgmt/kernelspec.py

```diff
@@ -109,12 +86,14 @@ def __str__(self):
 
 
 class KernelSpecManager:
-    def __init__(self, user_kernel_dir=None, kernel_dirs=None):
+    def __init__(self, user_kernel_dir=None, kernel_dirs=None, kernel_file=DEFAULT_KERNEL_FILE):
         super().__init__()
         self.user_kernel_dir = user_kernel_dir or self._user_kernel_dir_default()
         if kernel_dirs is None:
             kernel_dirs = self._kernel_dirs_default()
         self.kernel_dirs = kernel_dirs
+        self.kernel_file = kernel_file
+        self.log = get_app_logger()
```

Comment:
I was broadly trying to keep traitlets out of everything that wasn't already inheriting from HasTraits, which is why I was using the logging framework directly here. Probably doesn't matter much, though.

Reply:
Yes, I understand.  I think providers are going to want some notion of configurability as well, so I think keeping traitlets out of the picture is going to be difficult.  That said, I'm hoping this is the only occurrence at this level.

Reply:
My general argument is that configuration should happen at the application level, not the library level. Otherwise you end up with situations where the application wants to tell the library to do X, but the config is telling it to do Y, and someone's going to be disappointed.

I'm not purist about this, though. The docker and ssh providers I wrote as examples both read a config file - though arguably that's more defining kernels rather than configuring library behaviour. Not sure exactly where the distinction lies.